r_code/panel_estimators.R

```diff
@@ -0,0 +1,750 @@
+                                  pre_periods = 0,
+                                  post_periods = 20,
+                                  nnsize = NULL,
+                                  scale = 100,
+                                  period = 52,
+                                  constrained = T) {
+  # Separate the treatment time and the outcome process.
+  treatperiod <- y_df[1, 2]
+  y <- y_df[1:(treatperiod - 1), 1]
+  # Rescale the outcomes for control and treated.
+  y_con <- ct_mat / scale
+  nc <- ncol(y_con)
+  np <- nrow(y_con)
+  y_pre <- c(y / scale, rep(0, np - treatperiod + 1))
+
+  # tv is a linear weight vector. Periods before treatperiod - period use
+  # zero weights.
+  # nnsize neighbours are identified using weighted distances before the
+  # treatment period.
+  start_period <- max(1, (treatperiod - 1) - period - 1)
+  tv <- c(rep(0, start_period), seq_len(np - start_period))
+  tmc <- t(matrix(tv, nrow = nc, ncol = length(tv), byrow = TRUE))
+  wt_y_con <- sqrt(tmc) * y_con
+  wt_y_pre <- sqrt(tv) * y_pre
+
+  # If nnsize is NULL, the number of the neighbours are chosen to be close
+  # to the number of periods.
+  if (is.null(nnsize)) {
+    nnsize <- treatperiod
+  }
+  # Compute the distance between the time series and order them by distance.
+  y_r <- t(matrix(wt_y_pre, nrow = nc, ncol = np, byrow = TRUE))
+  e <-
+    (y_r[seq_len(treatperiod - 1), ] - wt_y_con[seq_len(treatperiod - 1), ])**2
+
+  es <- colSums(e)
+  es_order <- order(es)
+  # Combine the nearest neighbors and the treated unit, compute the predictions.
+  y_input <- cbind(y_con[, es_order[seq_len(min(nnsize, nc))]], y_pre)
+  pred <- .SingleSDIDPredict(t(y_input), treatperiod,
+    pre_periods = pre_periods,
+    post_periods = post_periods,
+    constrained = constrained
+  )
+  out <- pred * scale
+  return(out)
+}
+
+#' Compute the SDID estimates of the outcome series.
+#'
+#' @inheritParams .SDIDPredictionHelper
+#' @inheritParams EstimateCausalImpactSeries
+#'
+#' @return A long form tibble with the SDID method estimates of the
+#'    outcome variable (missing outcome imputation), as well as point and
+#'    percent effects.
+# TODO(alexdkellogg): Replace/Compare with the 2 way fixed effects, full panel
+#    version of SDID as in EQ 2.3 in Athey et al 2019.
+EstimateSDIDSeries <- function(data_full,
+                               id_var = "entry",
+                               time_var = "period",
+                               treat_indicator = "treatperiod_0",
+                               outcome_var = "target",
+                               counterfac_var = "counter_factual",
+                               pre_periods = 0,
+                               post_periods = 20,
+                               nnsize = NULL,
+                               scale = 100,
+                               period = 52,
+                               constrained = T) {
+  # Identify the treated and control ids.
```

Comment:
\# TODO(alexkellogg): Consider reducing the duplicated code also in EstimateSCMSeries.

Reply:
Yup, in all of them really. Thanks.